File: src/c/ch01/1.4/l0/demo_overflow_precision_beginner.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include "int_overflow_sim.h"
#include "float_precision.h"

int main(void) {
    puts("=== Integer overflow (8-bit) ===");
    printf("255 + 1 (unsigned, wrap) -> %u\n",
           add_unsigned_bits(255u, 1u, 8));                 /* 0 */
    printf("127 + 1 (signed, wrap)   -> %d\n",
           add_signed_bits(127, 1, 8));                     /* -128 */

    uint8_t r; bool cf, of;
    add_with_flags_8bit(200u, 100u, &r, &cf, &of);
    printf("200 + 100 -> result=%u (unsigned), CF=%s, OF=%s\n",
           (unsigned)r, cf ? "true" : "false", of ? "true" : "false");
    printf("Interpret result as signed: %d\n", to_signed_bits(r, 8));

    puts("\n=== Floating-point surprises ===");
    double x = 0.1 + 0.2;
    printf("0.1 + 0.2 = %.17g\n", x);                     /* 0.30000000000000004 */
    printf("Direct equality with 0.3? %s\n", (x == 0.3) ? "true" : "false");
    printf("Using epsilon: %s\n", almost_equal(x, 0.3, 1e-9) ? "true" : "false");

    puts("\nRepeat add 0.1 ten times:");
    double s = repeat_add(0.1, 10);
    printf("sum = %.17g, equal to 1.0? %s, almost_equal? %s\n",
           s, (s == 1.0) ? "true" : "false",
           almost_equal(s, 1.0, 1e-9) ? "true" : "false");

    puts("\nMix large and small:");
    double a = mix_large_small(1e16, 1.0);
    printf("1e16 + 1.0 = %.17g (often unchanged)\n", a);

    puts("\nNaive sum may drift slightly:");
    double nums[10];
    for (int i = 0; i < 10; ++i) nums[i] = 0.1;
    printf("sum_naive([0.1]*10) = %.17g\n", sum_naive(nums, 10));

    return 0;
}

File: archived/src/python/ch01/1.4/l0/demo_overflow_precision_beginner.py
from int_overflow_sim import add_unsigned, add_signed, add_with_flags_8bit, to_signed
from float_precision import almost_equal, repeat_add, mix_large_small, sum_list

def main():
    print("=== Integer overflow (8-bit) ===")
    print("255 + 1 (unsigned, wrap):", add_unsigned(255, 1, bits=8))  # 0
    print("127 + 1 (signed, wrap):  ", add_signed(127, 1, bits=8))    # -128

    r, cf, of = add_with_flags_8bit(200, 100)
    print(f"200 + 100 -> result={r} (unsigned), CF={cf}, OF={of}")
    print("Interpret result as signed:", to_signed(r, 8))

    print("\n=== Floating-point surprises ===")
    x = 0.1 + 0.2
    print("0.1 + 0.2 =", x)                  # 0.30000000000000004
    print("Direct equality with 0.3?", x == 0.3)
    print("Using epsilon:", almost_equal(x, 0.3))  # True

    print("\nRepeat add 0.1 ten times:")
    s = repeat_add(0.1, 10)
    print("sum =", s, "equal to 1.0?", s == 1.0, "almost_equal?", almost_equal(s, 1.0))

    print("\nMix large and small:")
    a = mix_large_small(1e16, 1.0)
    print("1e16 + 1.0 =", a, " (unchanged)")

    print("\nNaive sum may drift slightly:")
    nums = [0.1] * 10
    print("sum_list([0.1]*10) =", sum_list(nums))

if __name__ == "__main__":
    main()
